go/internal/feast/featurestore.go

```diff
@@ -0,0 +1,885 @@
+				featureName = parsedFeatureName[0]
+			} else {
+				featureName = parsedFeatureName[1]
+			}
+			requestedResultRowNames[featureName] = struct{}{}
+		}
+	}
+
+	featureViews := make([]*FeatureView, len(groupedRefs))
+	index := 0
+	for featureView := range groupedRefs {
+		featureViews[index] = featureView
+		index += 1
+	}
+
+	// TODO (Ly): This should return empty now
+	// Expect no ODFV or Request FV passed in GetOnlineFearuresRequest
+	neededRequestData, neededRequestFvFeatures, err := fs.getNeededRequestData(groupedOdfvRefs, groupedRequestFvRefs)
+	if err != nil {
+		return nil, err
+	}
+
+	joinKeyValues := make(map[string]*types.RepeatedValue)
+	requestDataFeatures := make(map[string]*types.RepeatedValue) // TODO (Ly): Should be empty now until ODFV and Request FV are supported
+	for entityName, vals := range entityProtos {
+		if _, ok := neededRequestFvFeatures[entityName]; ok {
+			requestedResultRowNames[entityName] = struct{}{}
+			requestDataFeatures[entityName] = vals
+		} else if _, ok = neededRequestData[entityName]; ok {
+			requestDataFeatures[entityName] = vals
+		} else {
+			if joinKey, ok := entityNameToJoinKeyMap[entityName]; !ok {
+				return nil, errors.New(fmt.Sprintf("entityNotFoundException: %s\n%v", entityName, entityNameToJoinKeyMap))
+			} else {
+				joinKeyValues[joinKey] = vals
+				requestedResultRowNames[joinKey] = struct{}{}
+			}
+		}
+
+	}
+
+	// TODO (Ly): Skip this validation since we're not supporting ODFV yet
+
+	// err = fs.ensureRequestedDataExist(neededRequestData, neededRequestFvFeatures, requestDataFeatures)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	onlineFeatureResponse := &serving.GetOnlineFeaturesResponse{Metadata: &serving.GetOnlineFeaturesResponseMetadata{
+		FeatureNames: &serving.FeatureList{Val: make([]string, 0)},
+	},
+		Results: make([]*serving.GetOnlineFeaturesResponse_FeatureVector, numRows),
+	}
+	// Allocate memory for each GetOnlineFeaturesResponse_FeatureVector
+	for index = 0; index < numRows; index++ {
+		onlineFeatureResponse.Results[index] = &serving.GetOnlineFeaturesResponse_FeatureVector{Values: make([]*types.Value, 0),
+			Statuses:        make([]serving.FieldStatus, 0),
+			EventTimestamps: make([]*timestamppb.Timestamp, 0),
+		}
+	}
+	// Merge joinKeyValues into requestDataFeatures
+	for entityName, vals := range joinKeyValues {
+		requestDataFeatures[entityName] = vals
+	}
+
+	fs.populateResultRowsFromColumnar(onlineFeatureResponse, requestDataFeatures)
+
+	entitylessCase := false
+
+	for _, featureView := range featureViews {
+		if _, ok := featureView.entities[DUMMY_ENTITY_NAME]; ok {
+			entitylessCase = true
+			break
+		}
+	}
+
+	if entitylessCase {
+		dummyEntityColumn := &types.RepeatedValue{Val: make([]*types.Value, numRows)}
+		for index := 0; index < numRows; index++ {
+			dummyEntityColumn.Val[index] = &DUMMY_ENTITY
+		}
+		joinKeyValues[DUMMY_ENTITY_ID] = dummyEntityColumn
+	}
+
+	for table, requestedFeatures := range groupedRefs {
```

Comment:
Since we rewriting this anyway, I think it's a good chance to finally fix this behaviour. We should group features by entities, not by feature views (also name `table` here is even more misleading), and call `readFromOnlineStore` once per entity not once per feature view. Hence, let storage layer to decide what storage model should be used and leave room for optimization.

Reply:
Hi Oleksii, so instead of querying a set of entity keys (serves as primary key) as a key in hash table (e.g: redis) which could uniquely identify a set of feature views and return those and only get the features from feature views we want, would your suggestion instead to query one entity at a time and retrieve all feature views that contains this entity key but not necessarily uniquely identifies a row in those feature views? If this is the case, we would have to change the implementation of materialization (currently subject to python sdk), which maybe out of scope for this PR? Do you have any suggestion on how we can initiate that change but in-scope for this PR?

Reply:
exactly